src/interpreter.rs

```diff
@@ -260,179 +257,52 @@ impl<T: Clone> Stack<T> {
-    if usize::from(sig[1]) != sig.len() - 3 {
-        return false;
-    };
-
-    // Extract the length of the R element.
-    let len_r = usize::from(sig[3]);
-
-    // Make sure the length of the S element is still inside the signature.
-    if 5 + len_r >= sig.len() {
-        return false;
-    };
-
-    // Extract the length of the S element.
-    let len_s = usize::from(sig[5 + len_r]);
-
-    // Verify that the length of the signature matches the sum of the length
-    // of the elements.
-    if len_r + len_s + 7 != sig.len() {
-        return false;
-    };
-
-    // Check whether the R element is an integer.
-    if sig[2] != 0x02 {
-        return false;
-    };
-
-    // Zero-length integers are not allowed for R.
-    if len_r == 0 {
-        return false;
-    };
-
-    // Negative numbers are not allowed for R.
-    if sig[4] & 0x80 != 0 {
-        return false;
-    };
-
-    // Null bytes at the start of R are not allowed, unless R would
-    // otherwise be interpreted as a negative number.
-    if len_r > 1 && sig[4] == 0x00 && sig[5] & 0x80 == 0 {
-        return false;
-    };
-
-    // Check whether the S element is an integer.
-    if sig[len_r + 4] != 0x02 {
-        return false;
-    };
-
-    // Zero-length integers are not allowed for S.
-    if len_s == 0 {
-        return false;
-    };
-
-    // Negative numbers are not allowed for S.
-    if sig[len_r + 6] & 0x80 != 0 {
-        return false;
-    };
-
-    // Null bytes at the start of S are not allowed, unless S would otherwise be
-    // interpreted as a negative number.
-    if len_s > 1 && sig[len_r + 6] == 0x00 && sig[len_r + 7] & 0x80 == 0 {
-        return false;
-    };
-
-    true
 }
 
-fn is_low_der_signature(vch_sig: &ValType) -> Result<bool, ScriptError> {
-    if !is_valid_signature_encoding(vch_sig) {
-        return set_error(ScriptError::SigDER);
-    };
-    // https://bitcoin.stackexchange.com/a/12556:
-    //     Also note that inside transaction signatures, an extra hashtype byte
-    //     follows the actual signature data.
-    let (_, vch_sig_copy) = vch_sig
-        .split_last()
-        .expect("`is_valid_signature_encoding` checks that the length is at least 9");
-    // If the S value is above the order of the curve divided by two, its
-    // complement modulo the order could have been used instead, which is
-    // one byte shorter when encoded correctly.
-    // FIXME: This can return `false` without setting an error, which is not the expectation of the
-    //        caller.
-    Ok(PubKey::check_low_s(vch_sig_copy))
+#[derive(Clone)]
+pub struct Signature {
+    sig: ecdsa::Signature,
+    sighash: HashType,
 }
 
-fn is_defined_hashtype_signature(vch_sig: &ValType) -> bool {
-    if vch_sig.is_empty() {
-        return false;
-    };
-
-    HashType::from_bits(i32::from(vch_sig[vch_sig.len() - 1]), true).is_ok()
+fn decode_signature(vch_sig_in: &[u8], is_strict: bool) -> Result<Option<Signature>, ScriptError> {
```

Comment:
This function needs documentation. The fact that `Ok(None)` is effectively a deferred error is not at all clear, and now that `decode_signature` has a return value (vs `check_signature_encoding`), it is important that the caller handle it correctly (which they may not do if they don't go look at the internals of this method each time they interact with its callsites).